File: Ruby/lib/civitas-ruby/jugador.rb
require_relative 'diario'
module Civitas
  class Jugador
    @@Casas_max = 4
    @@Casas_por_hotel = 4
    @@Hoteles_max = 4
    @@Paso_por_salida = 1000
    @@Precio_libertad = 200
    @@Saldo_inicial = 7500
    
    def initialize(nombre)
      @encarcelado = false
      @nombre = nombre
      @num_casilla_actual = 0
      @puede_comprar
      @saldo = @@Saldo_inicial
      @propiedades = []
      @salvoconducto = nil
    end
    
    attr_reader :num_casilla_carcel, :num_casilla_actual, :puede_comprar, :encarcelado
    attr_reader :nombre
    attr_reader :propiedades, :saldo
    
    attr_writer :saldo


    def cancelar_hipoteca(ip)
      result = false
      if encarcelado
        result
      else
        if existe_la_propiedad(ip)
          propiedad = @propiedades[ip]
          cantidad = propiedad.importe_cancelar_hipoteca
          if puedo_gastar(cantidad)
            result = propiedad.cancelar_hipoteca(self)
            if result
              Diario.instance.ocurre_evento("El jugador #{nombre} cancela la hipoteca de la propiedad #{ip}")
            end
          end
        end
      end
      result
    end

    def cantidad_casas_hoteles
      i = 0
      @propiedades.each { |propiedad| i += (propiedad.num_casas + propiedad.num_hoteles)}
      i
    end

    def comprar(titulo)
      result = false
      if @encarcelado
        result
      else
        if @puede_comprar
          precio = titulo.precio_compra
          if puedo_gastar(precio)
            result = titulo.comprar(self)
            if result
              @propiedades << titulo
              Diario.instance.ocurre_evento("El jugador #{@nombre} compra la propiedad #{titulo.nombre}")
            end
            @puede_comprar = false
          end
        end
      end
      result
    end

    def construir_casa(ip)
      result = false
      if encarcelado
        result
      else
        existe = existe_la_propiedad(ip)
        if existe
          propiedad = @propiedades[ip]
          puedo_casa = puedo_edificar_casa(propiedad)
          if puedo_casa
            result = propiedad.construir_casa(self)
            if result
              Diario.instance.ocurre_evento("El jugador #{@nombre} construye casa en la propiedad #{ip}")
            end
          end
        end
      end
      result
    end

    def construir_hotel(ip)
      result = false
      if @encarcelado
        result
      else
        if existe_la_propiedad(ip)
          propiedad = @propiedades[ip]
          puedo_hotel = puedo_edificar_hotel(propiedad)
          if puedo_hotel
            result = propiedad.construir_hotel(self)
            casas_hotel = @@Casas_por_hotel
            propiedad.derruir_casas(casas_hotel, self)
            Diario.instance.ocurre_evento("El jugador #{@nombre} construye hotel en la propiedad #{ip}")
          end
        end
      end
      result
    end

    def en_bancarrota
      @saldo < 0
    end

    def encarcelar(num_casilla_carcel)
      if debe_ser_encarcelado
        mover_a_casilla(num_casilla_carcel)
        @encarcelado = true
        Diario.instance.ocurre_evento("#{@nombre} ha sido encarcelado")
      end
      @encarcelado
    end

    def self.Casas_por_hotel
      @@Casas_por_hotel
    end

    def hipotecar(ip)
      result = false
      if @encarcelado
        result
      else
        if existe_la_propiedad(ip)
          propiedad = @propiedades[ip]
          result = propiedad.hipotecar(self)
        end
        if result
          Diario.instance.ocurre_evento("El jugador #{@nombre} hipoteca la propiedad #{ip}")
        end
      end
      result
    end

    def mover_a_casilla(num_casilla)
      if encarcelado
        false
      else
        @num_casilla_actual = num_casilla
        @puede_comprar = false
        Diario.instance.ocurre_evento("#{@nombre} se mueve a casilla #{num_casilla_actual}")
        true
      end
    end

    def obtener_salvoconducto(sorpresa)
      if encarcelado
        false
      else
        @salvoconducto = sorpresa
        true
      end
    end

    def paga(cantidad)
      @saldo -= cantidad
    end

    def paga_alquiler(cantidad)
      if encarcelado
        false
      else
        paga(cantidad)
      end
    end

    def paga_impuesto(cantidad)
      if encarcelado
        false
      else
        paga(cantidad)
      end
    end

    def pasa_por_salida
      @saldo += @@Paso_por_salida
      Diario.instance.ocurre_evento("#{@jugador} pasó por salida y cobró #{@@Paso_por_salida}")
      true
    end

    def puede_comprar_casilla
      if encarcelado
        @puede_comprar = false
      else
        @puede_comprar = true
      end
    end

    def recibe(cantidad)
      if encarcelado
        false
      else
        @saldo += cantidad
      end
    end

    def salir_carcel_pagando
      puts "Dejame pagar coño"
      if encarcelado && puede_salir_carcel_pagando
        paga(@@Precio_libertad)
        @encarcelado = false
        Diario.instance.ocurre_evento("#{@jugador} sale de la cárcel pagando")
        true
      else
        false
      end
    end

    def salir_carcel_tirando
      if Dado.instance.salgo_de_la_carcel
        @encarcelado = false
        Diario.instance.ocurre_evento("#{@jugador} sale de la cárcel tirando")
      end
      puts "#{Dado.instance.ultimo_resultado} != 5"
      @encarcelado
    end

    def tiene_algo_que_gestionar
      !@propiedades.empty?
    end

    def tiene_salvoconducto
      @salvoconducto != nil
    end

    def to_s
      s = "Nombre: #{@nombre}
      Casilla actual: #{@num_casilla_actual}
      Saldo: #{@saldo}
      Propiedades: #{@propiedades.map { |p| p.to_s }}
      #{@salvoconducto != nil ? "T" : "No t"}iene salvoconducto
      #{@puede_comprar ? "P" : "No p"}uede comprar
      #{@encarcelado ? "E" : "No e"}stá encarcelado"
    end

    def vender(ip)
      if encarcelado
        false
      else
        if existe_la_propiedad(ip)
          if propiedades[ip].vender(self)
            Diario.instance.ocurre_evento("#{@jugador} vende la propiedad #{@propiedades[ip].nombre}")
            propiedades.delete_at(ip)
            true
          else
            false
          end
        else
          false
        end
      end
    end

    def <=>(otro)
      otro.saldo <=> @saldo
    end

    private

    def existe_la_propiedad(ip)
      ip < @propiedades.size
    end

    def self.Casas_max
      @@Casas_max
    end

    def self.Hoteles_max
      @@Hoteles_max
    end

    def self.Precio_libertad
      @@Precio_libertad
    end

    def self.Paso_por_salida
      @@Paso_por_salida
    end

    def perder_salvoconducto
      @salvoconducto.usada
      @salvoconducto = nil
    end

    def puede_salir_carcel_pagando
      @saldo >= @@Precio_libertad
    end

    def puedo_edificar_casa(propiedad)
      puedo_casa = false
      precio = propiedad.precio_edificar
      if puedo_gastar(precio) && propiedad.num_casas < @@Casas_max
        puedo_casa = true
      end
      puedo_casa
    end

    def puedo_edificar_hotel(propiedad)
      puedo_hotel = false
      precio = propiedad.precio_edificar
      if puedo_gastar(precio) && propiedad.num_hoteles < @@Hoteles_max && propiedad.num_casas >= @@Casas_por_hotel
        puedo_hotel = true
      end
      puedo_hotel
    end

    def puedo_gastar(precio)
      if encarcelado
        false
      else
        @saldo >= precio
      end
    end

    protected 

    def debe_ser_encarcelado
      if @encarcelado
        false
      else
        if tiene_salvoconducto
          perder_salvoconducto
          Diario.instance.ocurre_evento("#{@nombre} utiliza salvoconducto")
          false
        else
          true
        end
      end
    end

    #attr_reader :propiedades, :saldo
    #attr_reader :nombre

    # Constructor de copia?
  end
end

File: Java/src/civitas/Jugador.java
package civitas;

import java.util.ArrayList;

public class Jugador implements Comparable<Jugador> {

  protected static final int CasasMax = 4;
  protected static final int CasasPorHotel = 4;
  protected Boolean encarcelado;
  protected static final int HotelesMax = 4;
  private String nombre;
  private int numCasillaActual;
  protected static final float PasoPorSalida = 1000;
  protected static final float PrecioLibertad = 200;
  private Boolean puedeComprar;
  private float saldo;
  private static final float SaldoInicial = 7500;
  private ArrayList<TituloPropiedad> propiedades;
  private Sorpresa salvoconducto;

  Boolean cancelarHipoteca(int ip) {
    Boolean result = false;
    if (encarcelado) {
      return result;
    }
    if (existeLaPropiedad(ip)) {
      TituloPropiedad propiedad = propiedades.get(ip);
      float cantidad = propiedad.getImporteCancelarHipoteca();
      Boolean puedoGastar = this.puedoGastar(cantidad);
      if (puedoGastar) {
        result = propiedad.cancelarHipoteca(this);
      }
      if (result) {
        Diario.getInstance().ocurreEvento(String.format("El jugador %s cancela la hipoteca de la propiedad %d", nombre, ip));
      }
    }
    return result;
  }

  int cantidadCasasHoteles() {
    int cantidad = 0;
    for (int i = 0; i<propiedades.size(); i++) {
      cantidad += propiedades.get(i).cantidadCasasHoteles();
    }
    return cantidad;
  }

  public int compareTo(Jugador otro) {
    return Float.compare(otro.saldo, saldo);
  }

  Boolean comprar (TituloPropiedad titulo) {
    Boolean result = false;
    if (encarcelado){
      return result;
    }
    if (puedeComprar) {
      float precio = titulo.getPrecioCompra();
      if (puedoGastar(precio)) {
        result = titulo.comprar(this);
        if (result) {
          propiedades.add(titulo);
          Diario.getInstance().ocurreEvento(String.format("El jugador %s compra la propiedad %s", nombre, titulo.getNombre()));
        }
      }
      puedeComprar = false;
    }
    return result;
  }

  Boolean construirCasa(int ip) {
    Boolean result = false;
    if (encarcelado) {
        return result;
    }
    Boolean existe = existeLaPropiedad(ip);
    if (existe) {
      TituloPropiedad propiedad = propiedades.get(ip);
      Boolean puedoEdificarCasa = puedoEdificarCasa(propiedad);
      float precio = propiedad.getPrecioEdificar();
      if (puedoGastar(precio) && propiedad.getNumCasas() < getCasasMax()) {
        puedoEdificarCasa = true;
      }
      if (puedoEdificarCasa) {
        result = propiedad.construirCasa(this);
        if (result) {
          Diario.getInstance().ocurreEvento(String.format("El jugador %s construye casa en la propiedad %d", nombre, ip));
        }
      }
    }
    return result;
  }

  Boolean construirHotel(int ip) {
    Boolean result = false;
    if (encarcelado) {
      return result;
    }
    if (existeLaPropiedad(ip)) {
      TituloPropiedad propiedad = propiedades.get(ip);
      Boolean puedoEdificarHotel = puedoEdificarHotel(propiedad);
      if (puedoEdificarHotel) {
        result = propiedad.construirHotel(this);
        int casasPorHotel = getCasasPorHotel();
        propiedad.derruirCasas(casasPorHotel, this);
      }
      Diario.getInstance().ocurreEvento(String.format("El jugador %s construye hotel en la propiedad %d", nombre, ip));
    }
    return result;
  }

  protected Boolean debeSerEncarcelado() {
    Boolean debe = false;
    if (!encarcelado) {
      if (tieneSalvoconducto()) {
        perderSalvoconducto();
        Diario.getInstance().ocurreEvento("El jugador " + nombre + " se libra de la cárcel");
      }
      else
        debe = true;
    }
    return debe;
  }

  Boolean enBancarrota() {
    return saldo < 0;
  }

  Boolean encarcelar(int numCasillaCarcel) {
    // System.err.println("Estoy en jugador.encarcelar. Carcel en "+numCasillaCarcel);
    if (debeSerEncarcelado()) {
      moverACasilla(numCasillaCarcel);
      encarcelado = true;
      Diario.getInstance().ocurreEvento("El jugador " + nombre + " se encarcela");
    }
    return encarcelado;
  }

  private Boolean existeLaPropiedad(int ip) {
    return ip < propiedades.size() && ip >= 0;
  }

  private int getCasasMax() {
    return CasasMax;
  }

  int getCasasPorHotel() {
    return CasasPorHotel;
  }

  private int getHotelesMax() {
    return HotelesMax;
  }

  protected String getNombre() {
    return nombre;
  }
  
  int getNumCasillaActual() {
    return numCasillaActual;
  }

  private float getPrecioLibertad() {
    return PrecioLibertad;
  }

  private float getPremioPasoSalida() {
    return PasoPorSalida;
  }

  protected ArrayList<TituloPropiedad> getPropiedades() {
    return propiedades;
  }

  Boolean getPuedeComprar() {
    return puedeComprar;
  }

  protected float getSaldo() {
    return saldo;
  }

  Boolean hipotecar(int ip) {
    Boolean result = false;
    if (encarcelado) {
      return result;
    }
    if (existeLaPropiedad(ip)) {
      TituloPropiedad propiedad = propiedades.get(ip);
      result = propiedad.hipotecar(this);
    }
    if (result) {
      Diario.getInstance().ocurreEvento(String.format("El jugador %s hipoteca la propiedad %d", nombre, ip));
    }
    return result;
  }

  public Boolean isEncarcelado() {
    return encarcelado;
  }

  Jugador(String n) {
    encarcelado = false;
    nombre = n;
    numCasillaActual = 0;
    saldo = SaldoInicial;
    puedeComprar = true;
    salvoconducto = null;
    propiedades = new ArrayList<TituloPropiedad>();
  } 

  protected Jugador(Jugador otro) {
    encarcelado = otro.encarcelado;
    nombre = otro.nombre;
    numCasillaActual = otro.numCasillaActual;
    saldo = otro.saldo;
    puedeComprar = otro.puedeComprar;
    salvoconducto = otro.salvoconducto;
    propiedades = new ArrayList<TituloPropiedad>();
    for (int i=0; i<otro.propiedades.size(); i++) {
      propiedades.add(otro.propiedades.get(i));
    }
  }

  Boolean modificarSaldo(float cantidad) {
    saldo += cantidad;
    Diario.getInstance().ocurreEvento("El jugador " + getNombre() + (cantidad >= 0 ? " recibe " : " paga ") + Math.abs(cantidad));
    return true;
  }

  Boolean moverACasilla(int numCasilla) {
    Boolean mueve = false;
    if (!encarcelado) {
      numCasillaActual = numCasilla;
      puedeComprar = false;
      Diario.getInstance().ocurreEvento("Se movió al jugador " + nombre + " a la casilla " + numCasilla);
      mueve = true;
    }
    return mueve;
  }

  Boolean obtenerSalvoconducto(Sorpresa sorpresa) {
    Boolean obtiene = false;
    if (!encarcelado) {
      salvoconducto = sorpresa;
      obtiene = true;
    }
    return obtiene;
  }

  Boolean paga(float cantidad) {
    return modificarSaldo(-cantidad);
  }

  Boolean pagaAlquiler(float cantidad) {
    Boolean paga = false;
    if (!encarcelado) {
      paga = paga(cantidad);
    }
    return paga;
  }

  Boolean pagaImpuesto(float cantidad) {
    Boolean paga = false;
    if (!encarcelado) {
      paga = paga(cantidad);
    }
    return paga;
  }
  
  Boolean pasaPorSalida() {
    modificarSaldo(getPremioPasoSalida());
    Diario.getInstance().ocurreEvento("Pasa por salida el jugador "+nombre);
    return true;
  }

  private void perderSalvoconducto() {
    salvoconducto.usada();
    salvoconducto = null;
  }

  Boolean puedeComprarCasilla() {
    if (encarcelado)
      puedeComprar = false;
    else
      puedeComprar = true;
    return puedeComprar;
  }

  private Boolean puedeSalirCarcelPagando() {
    return saldo >= getPrecioLibertad();
  }

  private Boolean puedoEdificarCasa(TituloPropiedad propiedad) {
    Boolean puede = false;
    if (!encarcelado) {
      Boolean esLaPropiedad = false;
      for (int i=0; i<propiedades.size() && !esLaPropiedad; i++) {
        esLaPropiedad = (propiedades.get(i).equals(propiedad));
      } 
      if (esLaPropiedad && propiedad.getNumCasas() < CasasMax && saldo >= propiedad.getPrecioEdificar()) {
        puede = true;
      }
    }
    return puede;
  }

  private Boolean puedoEdificarHotel(TituloPropiedad propiedad) {
    Boolean puedoEdificarHotel = false;
    float precio = propiedad.getPrecioEdificar();
    if (puedoGastar(precio)) {
      if (propiedad.getNumHoteles() < getHotelesMax()) {
        if (propiedad.getNumCasas() >= getCasasPorHotel()) {
          puedoEdificarHotel = true;
        }
      }
    }
    return puedoEdificarHotel;
  }

  private Boolean puedoGastar(float precio) {
    Boolean puede = false;
    if (!encarcelado) {
      puede = saldo >= precio;
    }
    return puede;
  }

  Boolean recibe(float cantidad) {
    Boolean recibir = false;
    if (!encarcelado) {
      recibir = modificarSaldo(cantidad);
    }
    return recibir;
  }

  Boolean salirCarcelPagando() {
    Boolean sale = false;
    if (encarcelado && puedeSalirCarcelPagando()) {
      paga(getPrecioLibertad());
      encarcelado = false;
      Diario.getInstance().ocurreEvento("Jugador "+nombre+" paga para salir de la cárcel");
      sale = true;
    }
    return sale;
  }

  Boolean salirCarcelTirando() {
    Boolean tirada = Dado.getInstance().salgoDeLaCarcel();
    if (tirada) {
      encarcelado = false;
      Diario.getInstance().ocurreEvento("El jugador "+nombre+" tira para salir de la cárcel");
    }
    return tirada;
  }

  Boolean tieneAlgoQueGestionar() {
    return !propiedades.isEmpty();
  }

  Boolean tieneSalvoconducto() {
    return (salvoconducto != null);
  }

  public String toString() {
    String s = String.format(
      "\n\n  Nombre: %s" +
      "\n  Casilla actual: %d" +
      (encarcelado ? "\n  S" : "\n  No s") + "e encuentra encarcelado" +
      "\n  Saldo: %f" +
      (tieneSalvoconducto() ? "\n  T" : "\n  No t") + "iene salvoconducto" +
      "\n  Propiedades: %s", nombre, numCasillaActual, saldo, propiedades.toString());
    return s;
  }

  Boolean vender(int ip) {
    Boolean puede = false;
    if (!encarcelado && existeLaPropiedad(ip)) {
      if (propiedades.get(ip).vender(this)) {
        propiedades.remove(ip);
        Diario.getInstance().ocurreEvento("Jugador "+nombre+" vende propiedad");
        puede = true;
      }
    }
    return puede;
  }

  public int getNumPropiedades() { // HOLA??????? 
    return propiedades.size();
  }
}